### DISCUSSION TITLE
Will Keda support activity triggers?

### DISCUSSION THREAD

**Discussion by justinmchase** [kedacore/keda#2832] [Q&A / Need Help]
On the [list of supported triggers](https://docs.microsoft.com/en-us/azure/azure-functions/functions-kubernetes-keda#supported-triggers-in-keda) in the Azure documentation Keda does not list `ActivityTrigger`, which is an essential part of Orchestrator functions.

![Screen Shot 2022-03-27 at 10 40 37 AM](https://user-images.githubusercontent.com/10974/160289244-7e3fd335-f0f9-4c37-a3f1-a0935ee0780b.png)

On the Azure [timeline for In-process -> isolated function transitioning](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/net-on-azure-functions-roadmap/ba-p/2197916), they state that later this year dotnet 7 will arrive which will have only isolated functions.

<img width="499" alt="dotnet-functions-roadmap-1" src="https://user-images.githubusercontent.com/10974/160288864-50d6deff-1a42-4790-b9bd-7db1b3d560b6.png">

In the Orchestrator function documentation they mention that Orchestrator functions will be fully _isolated_ in the intial dotnet 7 release. 

I am wondering, if there is any intention for Keda to support scaling for Activity triggers now or in the future?

I'm not entirely sure how its working, but it seems like the Orchestator / Activity functions are just using Azure storage queues and tables to do all of the orchestration.

Any info on this topic would be appreciated!

**Comment by zroubalik** [comment]
@tomkerkhove PTAL

**Comment by tomkerkhove** [comment]
This is a question for the Azure Functions team as we don't know the internals of that. I think it's best to log an issue on https://github.com/Azure/azure-functions-durable-extension

**Reply by justinmchase** [reply]
@tomkerkhove I'm not sure I understand, are they the ones maintaining the azure storage triggers? Or are you suggesting I would need to gather information from them on how to understand the orchestrator internals in order to translate it into a trigger here?

**Reply by tomkerkhove** [reply]
Microsoft maintains the storage scalers in KEDA but I'm not aware of how activities work and how best to scale them so it's best to ask that team for help but the above people should be able to help.

Long story short, it's not in KEDA today but it can be but we'll need to check. I'll pick this up after paternity leave (if I don't forget)

**Reply by justinmchase** [reply]
Ok thanks! I left a note below on what it looks like its doing but yeah if there was a more efficient way to get the counts of active activities that would be very useful.

Also, congrats on fatherhood!

**Comment by justinmchase** [comment]
It appears to be driven by Queues and Tables:

<img width="286" alt="Screen Shot 2022-04-04 at 8 24 28 AM" src="https://user-images.githubusercontent.com/10974/161554188-82d18506-f019-4a85-be53-4f2195c4a252.png">

But it seems like its using the same set of Queues for all activity functions which I believe could be across multiple deployments.

So it seems like what you'd really want to do would be to have an Activity Trigger type which essentially monitors events on the `TestHubNameHistory` table and look at `TaskScheduled` and `TaskCompleted` events, count the active tasks by `Name` then the trigger could essentially be something like:

```yaml
apiVersion: keda.sh/v1alpha1 
kind: ScaledObject
metadata:
  name: hello-scaler
spec:
  scaleTargetRef:
    name: hello-processor
  minReplicaCount: 1
  maxReplicaCount: 10
  triggers:
  - type: azure-activity
    metadata:
      activityName: 'Hello'
      count: 50
```

<img width="1316" alt="Screen Shot 2022-04-04 at 8 35 00 AM" src="https://user-images.githubusercontent.com/10974/161556328-638fb3fa-c107-45b5-b3c6-d4c8c109c383.png">

I don't know how you'd actually accomplish it... it doesn't seem like tables have aggregates (such as count) and since the state of the activity is two different rows and you'd have to iterate the table to get state for each activity on some kind of an interval... Sounds gross, but possible.

**Reply by cgillum** [reply]
It’s actually a bit simpler than this. Activity triggers are driven entirely by the singular “-workitem” queue. One could either reuse the existing Azure Storage queue scaler and target this queue or write a new one that specifically targets this queue. No need to consider anything in Tables.

If you wanted to get fancy, you could write a more complex scaler that monitored the control queues as well, which drive orchestrator and entity functions. However, you can probably benefit 80% off users by just focusing on the one work item queue.

**Reply by justinmchase** [reply]
Right, but if they're all using the same queue but I have multiple pods with different sets of functions, how could I scale them differently?
